[ENH] Add Subsequence Extraction Transformer to `sktime.transformations` module

wirrywoo opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->

Fixes #6901.

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Subsequence extraction transformer is a transformer created to extract subsequences of specified length that
meet some criterion with respect to an aggregate function. It addresses questions like the following:

- Given a sequence $(x_1, x_2, ..., x_n)$, find a $k$ - length subsequence such that the average value of $(x_j, x_{j+1}, ..., x_{j+k-1})$ is maximal.

#### Does your contribution introduce a new dependency? If yes, which one?

Nope.

<!--
Only relevant if you changed pyproject.toml.
We try to minimize dependencies in the core dependency set. There
are also further specific instructions to follow for soft dependencies.
See here for handling dependencies in sktime: https://www.sktime.net/en/latest/developer_guide/dependencies.html
-->

#### What should a reviewer concentrate their feedback on?

- [x] Agreed naming choices for parameters and class variables involved
- [x] Potential areas of improvements regarding optimizing code performance and code refactoring
- [x] Ensure transformed results are accurate
- [x] Descriptive documentation

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?

Nope.

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

Excited to add my first PR to sktime. Really looking forward to the reviews!

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

<!--
Thanks for contributing!
-->

fkiraly reviewed (changes requested):
Some questions about the design here.

* why is this a panel transformer? Do we need information about the other sequences for one given sequence?
* related: why are we padding at first? Should the algorithm not also work for unequal length series?
* currently, we are carrying out the aggregation in `fit` and remembering indices for `transform`, correct? Should we not instead carry it out in `transform`, i nthe first place?
* we can carry out minimal parameter validation in `fit`, or in `__init__`, to avoid the "else" and "try-except" blocks later.

Further, code formatting checks are faliing, I assume that precommit has not been fully set up by you - here's how to set them up locally: https://www.sktime.net/en/latest/developer_guide/coding_standards.html#setting-up-local-code-quality-checks

wirrywoo commented:
Responses to @fkiraly:

> * why is this a panel transformer? Do we need information about the other sequences for one given sequence?

Ahh, I was referencing `TruncationTransformer` thinking that the functionality would be similar as I'm trying to identify the correct indices to extract the subsequences, but now I realized why `TruncationTransformer` would be considered a panel transformer while `SubsequenceExtractionTransformer` wouldn't. Perhaps it should be classified as a Series-to-series transformers under Window-based series transforms instead? I made this change in [this commit](https://github.com/sktime/sktime/pull/6967/commits/9391fdcbb3897c6866aeed710f24ce0ce79c6b3f).

> * related: why are we padding at first? Should the algorithm not also work for unequal length series?
 
Removed in [this commit](https://github.com/sktime/sktime/pull/6967/commits/02d431559f84e56aeffb6472fae7231dfcd8b42f#diff-e49b4403cc788bf8afd38b0dc74c64903465d3ceb5f290e87e118045cb7f0d7eL77-L78). Initially I had it to pass some checks, but found an alternative solution without needing to use `PaddingTransformer`.

> * currently, we are carrying out the aggregation in `fit` and remembering indices for `transform`, correct? Should we not instead carry it out in `transform`, i nthe first place?

Initially, I defined "fitting" as trying to compute the indices needed to retrieve the correct subsequence from the sequence. Then, I reserved the "transform" step to actually retrieve the subsequence from the input data. Do you think this is a good way to design this transformer or can the fitting definition be improved here?

> * we can carry out minimal parameter validation in fit, or in __init__, to avoid the "else" and "try-except" blocks later.

Refactored in [this commit](https://github.com/sktime/sktime/pull/6967/commits/33e20d44454005dc642f34745086c20bc0f3c14c#diff-e49b4403cc788bf8afd38b0dc74c64903465d3ceb5f290e87e118045cb7f0d7eL81-L105).

> Further, code formatting checks are faliing, I assume that precommit has not been fully set up by you - here's how to set them up locally: https://www.sktime.net/en/latest/developer_guide/coding_standards.html#setting-up-local-code-quality-checks

Ahh, I forgot that I reconfigured my WSL instance to use Ubuntu 22.04 and forgot to reinstall the pre-commits. Thanks!

fkiraly commented:
> Do you think this is a good way to design this transformer or can the fitting definition be improved here?

In general, the series seen in `transform` can be different from the one seen in `fit`. So, should we not carry out the computation in `transform`? Not entirely sure though.

> Perhaps it should be classified as a Series-to-series transformer under Window-based series transforms instead?

Yes, series-to-series is the correct type in my opinion.

wirrywoo commented:
> In general, the series seen in `transform` can be different from the one seen in `fit`. So, should we not carry out the computation in `transform`? Not entirely sure though.

I see. It does make sense especially if we are using a fitted transformer for inference purposes, so after more thought, I agree with you. Changes are in [this commit](https://github.com/sktime/sktime/pull/6967/commits/71f768cc9db74b51610d23bcb65657b2fc2581f9).

Thanks!

fkiraly commented:
great, restarting the tests.

wirrywoo commented:
Apologies for the failed pre-commit. I believe I resolved it this time. I need to double check if I set my pre-commits correctly on my local machine.

Can you restart the tests when you get a chance? Thanks!

fkiraly commented:
(another go)

wirrywoo commented:
Running into build issue #6772, namely [Read the Docs build](https://readthedocs.org/projects/sktime/builds/25319195/) failed due to timeout.

fkiraly reviewed (approved):
Looks good to me!

Suggestions for improvement, but not blocking:

* we could allow an arbitrary series-to-primitives transformer (from `sktime`), or a function with specific signature, as the `aggregate` algorithm
* the docstring should say whether `subsequence_len` is in `iloc` or `loc` units
* the docstring should say what exactly the algorithm is - it is implied but never explicitly stated
* are there any literature references for the use of this algorithm that one should add? It is simple, but perhaps where it is used as a component is helpful.
* Perhaps different naming might be more convenient to the user? Shorten `subsequence_len`, and call `method` (not very specific), `selector` or `select_fun` or`select_method` or similar?

wirrywoo commented:
Some updates from the great recommendations provided by @fkiraly (Thanks!):

> - we could allow an arbitrary series-to-primitives transformer (from sktime), or a function with specific signature, as the aggregate algorithm

I was able to add [numpy function support](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R45-R46) for `aggregate_fn` but I was unable to integrate sktime's series-to-primitives transformer since these transformers return `pd.DataFrame`.

> - the docstring should say whether subsequence_len is in iloc or loc units

Addressed [here](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R43).

> - the docstring should say what exactly the algorithm is - it is implied but never explicitly stated

Addressed [here](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R29-R35). Is there a way to check if the LaTeX is correctly rendered and if the hyperlink reference to Wikipedia is also correctly formatted?

> - are there any literature references for the use of this algorithm that one should add? It is simple, but perhaps where it is used as a component is helpful.

Added [here](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R64-R67). It took me a while to find one but it turns out that this transformer generalizes the [maximum sum subarray problem](https://en.wikipedia.org/wiki/Maximum_subarray_problem) to using other aggregate functions and selector methods. Because of this, I changed the default values of the transformer [here](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R85).

> - Perhaps different naming might be more convenient to the user? Shorten subsequence_len, and call method (not very specific), selector or select_fun orselect_method or similar?

Changed `subsequence_len` to `subseq_len` and `method` to `selector ` throughout [this commit](https://github.com/sktime/sktime/pull/6967/commits/eef8b96b0d41a97e73dff459d9ab87336cdbbba5#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R85).

fkiraly commented:
ok, great! Already approved before, but before approving again, we need to make sure that the tests run.

fkiraly commented on sktime/transformations/series/subsequence_extraction.py:
```diff
@@ -0,0 +1,194 @@
+    1. :math:`A(x_{i}, \cdots, x_{i+k-1})` is maximal when `selector = 'max'`, and
+    2. :math:`A(x_{i}, \cdots, x_{i+k-1})` is minimal when `selector = 'min'`.
+
+    When `aggregate_fn = np.sum` and `selector = 'max'`, the problem degenerates to the
```
to ensure the docstring displays code fonts correctly, use *double backticks*, not single backticks for code. Single backticks are reserved for rst expressions such as `:math: `.

Also, minor point about this sentence, I would lead with the special case: "Maximum sum subarray is a special case and can be obtained by setting ..." - if this is an important subcase, I would also mention it further up.

fkiraly commented on sktime/transformations/series/subsequence_extraction.py:
```diff
@@ -0,0 +1,194 @@
+        "handles-missing-data": False,
+    }
+
+    def __init__(self, subseq_len, aggregate_fn=np.sum, kwargs={}, selector="max"):
```
we must avoid functions and mutable defaults like the dict.

The pattern to deal with this is using `None`, and internally a private attribute `self._aggregate_fn` (since the attributes of same name should always be identical to init args).

fkiraly commented on sktime/transformations/series/subsequence_extraction.py:
```diff
@@ -0,0 +1,194 @@
+            instance.
+            ``create_test_instance`` uses the first (or only) dictionary in ``params``
+        """
+        params = [
```
we should also add one param set that has only defaults, except for `subseq_len` which needs to be set.

fkiraly commented on sktime/transformations/series/subsequence_extraction.py:
```diff
@@ -0,0 +1,194 @@
+        parameter_set : str, default="default"
+            Name of the set of test parameters to return, for use in tests. If no
+            special parameters are defined for a value, will return ``"default"`` set.
+
```
remove one newline

fkiraly reviewed (changes requested):
Only some small change requests remaining, see above.

wirrywoo commented:
Thanks for the review @fkiraly. Hopefully this will be the last batch of code updates. 😄 

> to ensure the docstring displays code fonts correctly, use double backticks, not single backticks for code. Single backticks are reserved for rst expressions such as `:math:` .
>
> Also, minor point about this sentence, I would lead with the special case: "Maximum sum subarray is a special case and can be obtained by setting ..." - if this is an important subcase, I would also mention it further up.

Addressed [here](https://github.com/sktime/sktime/pull/6967/commits/6d2c976c0ed3e06325aff1f9658525dee060cd76#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4L29-R39). 

> we must avoid functions and mutable defaults like the dict.
> 
> The pattern to deal with this is using `None`, and internally a private attribute `self._aggregate_fn` (since the attributes of same name should always be identical to init args).

Addressed [here](https://github.com/sktime/sktime/pull/6967/commits/6d2c976c0ed3e06325aff1f9658525dee060cd76#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4L85-R90). Similar to `func_fitter.py` [[code here](https://github.com/sktime/sktime/blob/main/sktime/param_est/compose/_func_fitter.py)], I made `aggregate_fn` a required argument rather than enforcing a default callable value. I think this is a better design as there's really no reason to actually have a default callable based on the goals of the transformer. I also defaulted `kwargs = None`.

> we should also add one param set that has only defaults, except for `subseq_len` which needs to be set.

Because I decided to make `aggregate_fn` a required parameter, I added a third case when all required parameters are defined [here](https://github.com/sktime/sktime/pull/6967/commits/6d2c976c0ed3e06325aff1f9658525dee060cd76#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4R193).


> remove one newline

Removed [here](https://github.com/sktime/sktime/pull/6967/commits/6d2c976c0ed3e06325aff1f9658525dee060cd76#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4L175).

fkiraly commented on sktime/transformations/series/subsequence_extraction.py:
```diff
@@ -0,0 +1,195 @@
+
+    Parameters
+    ----------
+    aggregate_fn : np._core._multiarray_umath._ArrayFunctionDispatcher
```
I think most users will not know what this means, I would suggest providing a more general signature (e.g., 2D array -> scalar), or giving a few common examples.

fkiraly reviewed (approved):
Looks good! left one non-blocking comment about documenting the options.

wirrywoo commented:
Thanks @fkiraly !

> I think most users will not know what this means, I would suggest providing a more general signature (e.g., 2D array -> scalar), or giving a few common examples.

Agreed. Updated [here](https://github.com/sktime/sktime/pull/6967/commits/83e37f7007d23c637a1fbed06aa75c4fcca12fda#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4L43-R45). I also updated the example with the required parameters as well [here](https://github.com/sktime/sktime/pull/6967/commits/83e37f7007d23c637a1fbed06aa75c4fcca12fda#diff-e573b0141647b79c420365e1df0291f4c8ee3105df3c9a766d6bac3cc977a7e4L61-R63).

wirrywoo commented:
Addressed merge conflicts in .all-contributorsrc [here](https://github.com/sktime/sktime/pull/6967/commits/d0d4912ff0e09096d45ab871055d9db1b292df8f).
